### PR TITLE
added edge add remove augmentation

### PR DESCRIPTION
**Pull request by chris-soelistyo**
- Added augmentations: add edge and remove edge
- Added some tests for it

### CODE REVIEW

File: grace/utils/augment_graph.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+from grace.base import GraphAttrs, Annotation
+
+
+class RandomEdgeAdditionAndRemoval:
+    """Rotate the image and graph in tandem; i.e., the graph x-y coordinates
+    will be transformed to reflect the image rotation.
+
+    Accepts an image stack of size (C,W,H) or (B,C,W,H) and a dictionary
+    which includes the graph object.
```
Comment (quantumjot) [inline grace/utils/augment_graph.py:15]: This needs updating

File: grace/utils/augment_graph.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+    Parameters
+    ----------
+    n_add : int
+        Number of new edges to add
+    n_remove : Tuple[int]
+        Number of existing edges to remove
```
Comment (quantumjot) [inline grace/utils/augment_graph.py:22]: Perhaps these should be fractional, given that graphs could be of any size?

File: grace/utils/augment_graph.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        self.n_remove = n_remove
+
+    def __call__(
+        self, x: torch.Tensor, graph: Dict[Union[str, GraphAttrs], Any]
+    ) -> Tuple[torch.Tensor, dict]:
+        edges_list = list(graph["graph"].edges)
+        max_node = graph["graph"].number_of_nodes()
+        edges_to_add = [
+            tuple(self.rng.integers(low=0, high=max_node, size=(2,)))
+            + ({GraphAttrs.EDGE_GROUND_TRUTH: Annotation.UNKNOWN},)
```
Comment (quantumjot) [inline grace/utils/augment_graph.py:44]: One option here we could have is to randomly assign a label to the edge, something like:

```python
edge_label = rng.choice(Annotation)
```

File: tests/test_augmentation.py [hunk 4]
```diff
@@ -179,3 +180,60 @@ def test_augment_rotate_image_and_graph(n, rot_angle, rot_center):
...
+        (0, 0),
+        (0, 1),
+        (1, 1),
+        (1, 2),
+        (4, 0),
+    ],
+)
+class TestAugmentGraphEdgeAdditionRemoval:
+    @pytest.fixture
+    def outputs(self, n_add, n_remove, default_rng):
```
Comment (quantumjot) [inline tests/test_augmentation.py:197]: This is pretty neat - I hadn't seen this way of doing things before

File: tests/test_augmentation.py [hunk 5]
```diff
@@ -179,3 +180,60 @@ def test_augment_rotate_image_and_graph(n, rot_angle, rot_center):
...
+        self, n_add, n_remove, default_rng, outputs
+    ):
+        add_edges = RandomEdgeAdditionAndRemoval(n_add, 0, default_rng)
+        remove_edges = RandomEdgeAdditionAndRemoval(0, n_remove, default_rng)
+        num_edges_init = outputs["graph"].number_of_edges()
+
+        image, target_added = add_edges(
+            outputs["image"], {"graph": outputs["graph"]}
+        )
+        num_edges_added = target_added["graph"].number_of_edges()
```
Comment (quantumjot) [inline tests/test_augmentation.py:231]: I guess this is technically `num_edges_augmented_add`?

File: tests/test_augmentation.py [hunk 6]
```diff
@@ -179,3 +180,60 @@ def test_augment_rotate_image_and_graph(n, rot_angle, rot_center):
...
+
+        image, target_removed = remove_edges(image, target_added)
+        num_edges_removed = target_removed["graph"].number_of_edges()
```
Comment (quantumjot) [inline tests/test_augmentation.py:234]: Similar here - I find the variable name to be a little confusing

File: grace/utils/augment_graph.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        n_add: int = 10,
+        n_remove: int = 10,
+        rng: np.random.Generator = np.random.default_rng(),
+    ):
+        self.rng = rng
+        self.n_add = n_add
+        self.n_remove = n_remove
+
+    def __call__(
+        self, x: torch.Tensor, graph: Dict[Union[str, GraphAttrs], Any]
```
Comment (quantumjot) [inline grace/utils/augment_graph.py:38]: I think this type annotation can be simplified, given that you're passing the target dictionary:

```suggestion
        self, x: torch.Tensor, graph: Dict[str, Any]
```

**Review by quantumjot — CHANGES_REQUESTED** [review]
A few changes, but otherwise looks great.

Might want to check that the dataloader can successfully apply both image and graph augmentations as part of the transform too.

File: grace/utils/augment_graph.py (hunk 3 above)
Comment (quantumjot) [reply grace/utils/augment_graph.py:44]: Although maybe it's good to leave it as is - I'm not sure what's best TBH

Comment (mooniean) [reply grace/utils/augment_graph.py:44]: ideally, could the edge annotation be a sort of average of the ones close to where the new edge is?

Comment (quantumjot) [reply grace/utils/augment_graph.py:44]: Again, I might be wrong here, but I feel like having it be random, adds to the augmentation aspect

Comment (chris-soelistyo) [reply grace/utils/augment_graph.py]: @mooniean @quantumjot I've implemented something along these lines with the `find_average_annotation ` function, let me know if this looks alright.

Comment (quantumjot) [reply grace/utils/augment_graph.py]: I think this should be optional. The default should be either random annotation or `UNKNOWN`

**Comment by chris-soelistyo** [comment]
@quantumjot Added three options for annotation assignment.

File: grace/utils/augment_graph.py [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+
+    if num_positive > num_negative:
+        return Annotation.TRUE_POSITIVE
+    elif num_negative > num_positive:
+        return Annotation.TRUE_NEGATIVE
+    else:
+        return Annotation.UNKNOWN
+
+
+ANNOTATION_MODES = {"random", "unknown", "average"}
```
Comment (quantumjot) [inline grace/utils/augment_graph.py:45]: For consistency I would suggest making this an Enum (like `Annotation` or `GraphAttrs`)

File: grace/utils/augment_graph.py [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+        rng: np.random.Generator = np.random.default_rng(),
+        annotation_mode: str = "random",
+    ):
+        self.rng = rng
+        self.n_add = n_add
+        self.n_remove = n_remove
+
+        if annotation_mode not in ANNOTATION_MODES:
+            raise ValueError(
+                "annotation_mode must be one of %r." % ANNOTATION_MODES
```
Comment (quantumjot) [inline grace/utils/augment_graph.py:82]: f-string?

File: grace/utils/augment_graph.py [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+            )
+
+        if annotation_mode == "random":
+            self.assign_annotation = lambda edge, graph: self.rng.choice(
+                Annotation
+            )
+        elif annotation_mode == "unknown":
```
Comment (quantumjot) [inline grace/utils/augment_graph.py:89]: I'd make this the `else`, i.e. catch-all case

**Review by quantumjot — APPROVED** [review]
Looks great! Again, a few suggestions just for consistency

**Pull request merged**